### ISSUE TITLE
Some ASCII symbols got wrong encoding in incoming chat/messages senders' number/row

### ISSUE THREAD

**Issue by osering** [juha-h/baresip-studio#365]
Asterisk/freePBX @ chan_dongle is used for SMS/USSD manipulations (sending to softphone).

There is obviously a problem with character encoding in incoming chat (for example, Linphone do it right).
symbol "^" is displayed as "%5e"
symbol ">" is displayed as "%3e"
Most probably bunch of others also are crippled.
symbol "/" is displayed as "/"

Dialplan row generating that::
exten => ussd,n,MessageSend(pjsip:trunk_name/sip:name@sip.antisip.com,${CALLERID(num)}^${DONGLENAME:6})

Result (senders name/row) IS "ussd%5e1" "ussd%3e1" and "ussd/1"
Result (senders name/row) SHOULD BE "ussd^1" "ussd>1" and "ussd/1"

> I just tested by sending this message from Linux baresip cli app to Android baresip app:
> 
> ```
> Here comes some characters: ^ > /
> ```
> 
> and it appeared exactly the same on baresip Android app.

The problem is with character in "senders name"/From header, not in "message body".

**Comment by juha-h** [comment]
Oskars G writes:

> Asterisk/freePBX @ chan_dongle is used for SMS/USSD manipulations (sending to softphone).
> 
> There is obviously a problem with character encoding in incoming chat (for example, Linphone do it right).
> symbol "^" is displayed as "%5e"
> symbol ">" is displayed as "%3e"
> Most probably bunch of others also are crippled.
> symbol "/" is displayed as "/"
> 
> Dialplan row generating that::
> exten => ***@***.***,${CALLERID(num)}^${DONGLENAME:6})
> 
> Result (senders name/row) IS "ussd%5e1" "ussd%3e1" and "ussd/1"
> Result (senders name/row) SHOULD BE "ussd^1" "ussd>1" and "ussd/1"

Turn on SIP trace and show logcat of the incoming MESSAGE.

**Comment by juha-h** [comment]
Oskars G writes:

> Asterisk/freePBX @ chan_dongle is used for SMS/USSD manipulations
> (sending to softphone).

I don't have that setup to test with.

> There is obviously a problem with character encoding in incoming chat (for example, Linphone do it right).
> symbol "^" is displayed as "%5e"
> symbol ">" is displayed as "%3e"
> Most probably bunch of others also are crippled.
> symbol "/" is displayed as "/"
> 
> Dialplan row generating that::
> exten => ***@***.***,${CALLERID(num)}^${DONGLENAME:6})

> Result (senders name/row) IS "ussd%5e1" "ussd%3e1" and "ussd/1"
> Result (senders name/row) SHOULD BE "ussd^1" "ussd>1" and "ussd/1"

As I already wrote, please provide debug
https://github.com/juha-h/baresip-studio/wiki/Debugging of the incoming MESSAGE
request.

**Comment by juha-h** [comment]
I just tested by sending this message from Linux baresip cli app to Android baresip app:
```
Here comes some characters: ^ > /
```
and it appeared exactly the same on baresip Android app.

**Comment by juha-h** [comment]
Most likely charset of the MESSAGE body is something else than ASCII.  So I need to see the MESSAGE.  If I send the same "Here comes ..." message from Linphone Android app to baresip Android app, the characters are correctly shown.

**Comment by osering** [comment]
05-05 12:02:10.130 15052   942 D Baresip Lib: 12:02:10.128#
05-05 12:02:10.130 15052   941 D Baresip Lib: got message 'Hei, Tavs kredita atlikums ir 0,00 EUR. Vari to izmantot lidz 30/05/2023. Tava Bite' from peer 'sip:ussd%3e1@olo-tst'
05-05 12:02:10.130 15052   941 D Baresip Lib: sending message -5476376636986701680/sip:ussd%3e1@olo-tst/Hei, Tavs kredita atlikums ir 0,00 EUR. Vari to izmantot lidz 30/05/2023. Tava Bite
05-05 12:02:10.131 15052   941 D Baresip : Message event for -5476376636986701680 from sip:ussd%3e1@olo-tst at 1683277330130
05-05 12:02:10.131 15052   942 D Baresip Lib: TLS 94.23.17.185:9091 -> 192.168.8.109:39692
05-05 12:02:10.131 15052   942 D Baresip Lib: MESSAGE sip:receiver-0xb4000076b828c890@192.168.8.109:43287;transport=tls SIP/2.0
05-05 12:02:10.131 15052   942 D Baresip Lib: V
05-05 12:02:10.131 15052   942 D Baresip Lib: ia: SIP/2.0/TLS 94.23.17.185:9091;branch=z9hG4bKe0e2.504f73dc26072b6d93aa6f98733d83b2.1
05-05 12:02:10.131 15052   942 D Baresip Lib: Via: SIP/2.0/UDP 192.168.8.168:5060;re
05-05 12:02:10.131 15052   942 D Baresip Lib: ceived=77.219.8.201;rport=39548;branch=z9hG4bKPj94373e82-113e-4a43-8739-3f220259113f
05-05 12:02:10.131 15052   942 D Baresip Lib: From: <sip:ussd%3e1@olo-tst>;tag=48acfcd7
05-05 12:02:10.131 15052   942 D Baresip Lib: -fac6-42c4-885f-4844ef6dad80
05-05 12:02:10.131 15052   942 D Baresip Lib: To: <sip:receiver@sip.antisip.com>
05-05 12:02:10.131 15052   942 D Baresip Lib: Contact: <sip:sender-antisip@192.168.8.168:5060;alias=77.219.8
05-05 12:02:10.131 15052   942 D Baresip Lib: .201~39548~1>
05-05 12:02:10.131 15052   942 D Baresip Lib: Call-ID: dfb6696a-9a9f-45ef-bda4-a731b87819c0
05-05 12:02:10.131 15052   942 D Baresip Lib: CSeq: 29116 MESSAGE
05-05 12:02:10.131 15052   942 D Baresip Lib: Max-Forwards: 69
05-05 12:02:10.131 15052   942 D Baresip Lib: User-Agent: IncrediblePBX-
05-05 12:02:10.131 15052   942 D Baresip Lib: 16.0.33(19.0)
05-05 12:02:10.131 15052   942 D Baresip Lib: Content-Type: text/plain
05-05 12:02:10.131 15052   942 D Baresip Lib: Content-Length:    83
05-05 12:02:10.131 15052   942 D Baresip Lib: 
05-05 12:02:10.131 15052   942 D Baresip Lib: Hei, Tavs kredita atlikums ir 0,00 EUR. Vari to izmantot lidz
05-05 12:02:10.131 15052   942 D Baresip Lib:  30/05/2023. Tava Bite
05-05 12:02:10.131 15052   942 D Baresip Lib: receiver@sip.antisip.com: selected for oskars-0xb4000076b828c890
05-05 12:02:10.137 15052   941 D Baresip : Saved 71 messages
05-05 12:02:10.142 15052   942 D Baresip Lib: 12:02:10.142#
05-05 12:02:10.143 15052   942 D Baresip Lib: TLS 192.168.8.109:39692 -> 94.23.17.185:9091
05-05 12:02:10.143 15052   942 D Baresip Lib: SIP/2.0 200 OK
05-05 12:02:10.143 15052   942 D Baresip Lib: Via: SIP/2.0/TLS 94.23.17.185:9091;branch=z9hG4bKe0e2.504f73dc26072b6d93aa6f98733d83b2.1
05-05 12:02:10.143 15052   942 D Baresip Lib: Via: SIP/2.0/UDP 192.168.8.168:5060;received=77.219.8.201;rport=39548;branch=z9hG4bKPj94373e82-113e-4a43-8739-3f220259113f
05-05 12:02:10.143 15052   942 D Baresip Lib: From: <sip:ussd%3e1@olo-tst>;tag=48acfcd7-fac6-42c4-885f-4844ef6dad80
05-05 12:02:10.143 15052   942 D Baresip Lib: To: <sip:receiver@sip.antisip.com>;tag=8eb1ffa4a4ad7546
05-05 12:02:10.143 15052   942 D Baresip Lib: Call-ID: dfb6696a-9a9f-45ef-bda4-a731b87819c0
05-05 12:02:10.143 15052   942 D Baresip Lib: CSeq: 29116 MESSAGE
05-05 12:02:10.143 15052   942 D Baresip Lib: Server: baresip v3.0.0 (aarch64/Android)
05-05 12:02:10.143 15052   942 D Baresip Lib: Content-Length: 0
 
In second row: got message 'XXX' from peer 'sip:ussd%3e1@olo-tst'
this "%3e" instead of ">"

**Comment by juha-h** [comment]
OK, so the issue is that characters, such as >, that must be escaped when they appear in SIP URI userpart are not un-escaped when baresip app shows them.  Thanks for reporting, I'll take a look.

**Comment by juha-h** [comment]
After giving some thought to this issue, my conclusion is that it is not good idea to try to un-escape escaped characters in SIP URI userparts.  The reason is that those characters may not be printable characters at all (< %20) or may be non-ASCII characters (> %7f).

In general,  I don't think it is good practice to use characters that need to be escaped in SIP URI userparts.  I tried to create in Linphone account `lin<1@test.tutpro.com` and got `Error: failed to create account object`.  Same with ` lin%3e1@test.tutpro.com`.  So I don't see much point in displaying URIs that the app itself does not allow creating.

**Comment by juha-h** [comment]
If you want to avoid seeing the escaped characters, try creating a contact for URI `sip:ussd%3e1@olo-tst`.  Then you should see the contact name instead of the URI.

**Comment by juha-h** [comment]
I'll close this. See https://github.com/juha-h/baresip-studio/issues/365#issuecomment-1536365123.

**Comment by osering** [comment]
Why there should be forbidden escape characters from standard ASCII table (from 32 to 126(decimal)) used in standard SIP messaging? These are standard (not special) characters available on every qwerty keyboard and widely used, so is there reason should someone need hacking to get them shown in a way they are initially been sent?
No sure what is a use of non-ASCII non -printable characters to be translated to smth. like %3f here...

**Comment by juha-h** [comment]
See https://www.rfc-editor.org/rfc/rfc3261#section-25.1.  It does not place any limits on what the escaped characters can be.  There is even this sentence in the RFC:
```
Excluded US-ASCII characters ([RFC 2396](https://www.rfc-editor.org/rfc/rfc2396) [[5](https://www.rfc-editor.org/rfc/rfc3261#ref-5)]), such as space and control characters and characters used as URI delimiters, also MUST be escaped.  URIs MUST NOT contain unescaped space and control characters.
```
So the URI can contain non-printable control characters if they are escaped.

My advice is, don't use such characters in URI userpart that would need to be escaped or add a contact for such URIs.

**Comment by osering** [comment]
1. I am not saying there is obvious deviation from the standards. Just some other implementations, including opensource (code can be inspected) Linphone seems to be more user friendly (and proves, that it do not brake the system) and do not impose some unnecessary restrictions (to use some hacking/workaround to overcome those) on use of base ASCII standard printable characters. 

2. According to RFC2396: Uniform Resource Identifier (URI) is a compact string of characters  for identifying an abstract or physical resource.

3. In given usecase > ^ / are used (in Asterisk) as delimiters for the purpose separating senders_address and gateway_id which delivered message from sender (in given case "ussd>1" means given ussd message was received through gateway #1), so it together forms new Resource (according to RFC 2396 section 1.1.). Therefore banning characters ^ > | is not required.

4. Could not find in RFC 3261(SIP) Section 19.1., including 19.1.2 "Character Escaping Requirements", that Senders_address for the purpose of MessageSend is defined as SIP URI and should follow "escape rules".

5. SIP URI is formed by the sender. In this usecase Baresip-studio acts as receiver, not the sender (Request maker) who forms the senders_name according to some RFCs/standards and the message already reached destination point and has successfully traveled through all the servers (which is the cause for the escaping some characters).
So at this final stage (already at final receiver) starting to re-code printable characters to smth. seems unreasonable and not in line with requirements of RFCs quoted before.

**Comment by juha-h** [comment]
In your MESSAGE, escaped character `>` exists in From URI:
```
05-05 12:02:10.131 15052 942 D Baresip Lib: From: sip:ussd%3e1@olo-tst;tag=48acfcd7
```
According to RFC3261 section https://www.rfc-editor.org/rfc/rfc3261#section-25.1 From URI is SIP URI and SIP URI can contain `user` which is defined like this:
```
user             =  1*( unreserved / escaped / user-unreserved )
```
and `escaped` in defined like this:
```
 escaped     =  "%" HEXDIG HEXDIG
```
Thus any escaped characters can be includes in user, e.g., control characters.  Therefore baresip cannot decode escaped characters.

It is not only matter of From URIs, since when a message is received, it is also possible to send a message back or call the sender of the message.   And when that is done, characters that are invalid in Request URI and To URI would need to be escaped.

By the way, I just tried and it was not possible in Linphone to define an account where ^ or %3e character exists in its userpart.

**Comment by osering** [comment]
Arguments about "From URI" is reasonable and acceptable.

I see place for discussion concerning, at which stage escaped characters are to be checked/transformed /not allowed - at senders' end (when encoding/creating) or at receivers' end (when decoding). Logically it would be senders' responsibility. Not the receivers (destination client software) which for the sake of clean standard and possible servers' risks punish (by fixing senders errors) receiver converting already received (maybe, syntactically not corresponding RFC) From header characters and spoiling user experience by such "purism".

Nothing guarantees that every message sender is such that can be messaged back (with standard SIP setup), including such SIP infrastructure elements as SIP providers' (callcentric, sipgate e.t.c) bridges from<->to PSTN/mobile carrier (different implementations) and USSD (mobile service codes resent through PJSIP).

What about possibility to define account in Linphone (with such characters), it is different story, as it afterwards serves both purposes: show up in sending and receiving usecases. But here is discussion only about how to handle received escape characters, not that Client software should enable sending such characters. Asterisk as intermediary is totally another story and nobody can prohibit anyone to include one or another character in dialplans. Asterisk let you include such additional information in PJSIP headers, including custom headers //for example: `exten => addheader,1,Set(PJSIP_HEADER(add,X-MyHeader)=myvalue`)//, but clients' software still is not so advanced to be able to extract this header information from user defined custom headers and display in a custom way. This would be the best vector to follow and trying to mimic this info (like in this case) in Headers "From" field (which is standard and obligatory) is kind of workaround to overcome deficiencies (lack of customization options)  in SIP softphones (clients).
**Manipulation (read/write) freedom with custom headers would be super-great feature in Baresip-studio or any other opensource softphone on Linux/Android!**

![linphone-ussd](https://user-images.githubusercontent.com/62410690/236701370-bfb3a8d5-1bff-43c7-a8cb-941ae19bbb36.jpg)
Attached screenshot from Linphone, which received in parallel with Baresip-studio the same messages  (with "sensitive" characters in Asterisks' synthetically designed "From" header). P.S. It also strips domain part, which is not really expected feature...

**Comment by juha-h** [comment]
If From URI hostpart matches baresip account's hostpart, baresip shows
the message or call coming from the userpart.  If userpart contains @
character (%40), user would see the message or call coming from, for
example ***@***.***, which would be very confusing. This is yet another
example, why I'm not going to show un-escaped userpart to the user.  If
you need that, stick to Linphone.

**Comment by osering** [comment]
Just showing common practice in SIP clients world and giving some reasoning (about stages where strict standards control would be rational) to which I did not see any counterarguments. Maybe it could be implemented like in bash scripts, where you put "/" before some service characters to show them as ordinary readable characters
If custom headers could be implemented, it would massively overweight this feature.
Why confusing? Knowing how SIP URIs are baked up in Asterisk or other PBXes, such reaction could not be considered reasonable. On the other hand, if the user has zero knowledge/interest in SIP (protocol), why he/she would care or be surprised (more than seeing and trying to understand escaped character and RFCxxxx)?
I tried "@" in this usecase as well. I this case Baresip-studio is not converting (to %40), it is cutting off the second one, so right now there is no consistency anyway.

Thanks' for suggestion ;)

P.S. I highly evaluate your initiative and work on this great opensource project!
Just trying to help to make it better, not hinder.

**Comment by juha-h** [comment]
Oskars G writes:

> I tried "@" in this usecase as well. I this case Baresip-studio is not
> converting (to %40), it is cutting off the second one, so right now
> there is no consistency anyway.

Can you give more details about this test?  Looks like there is a bug
that needs to be fixed.

**Comment by osering** [comment]
What concerning "@", it is the same test, just instead of ">" I tried "@", so the URI could potentially look like "ussd@1@olo-tst". It delivered just "ussd@1" part. Have not tried wit more than one "@".
All these experiments was done by simple From header, containing just simple/compact URI a@b

Strictly according to RFC3261 (SIP), Section 20.20 "From", which refers to Section 20.10 "Contact", URI can be also like this 
From: "A. G. Bell" <sip:agb@bell-telephone.com>".

I formed it 2 ways (with UTF-8 characters (in Latvian you need them):
1) From: "dongle1 aāuū" <ussd-1@olo-tst>" (in practice you not always prefix with sip:)
2) From: "dongle1 aāuū" <sip:ussd-1@olo-tst>"
* domain @olo-tst was attached by FreePBX PJSIP trunk (Advanced) setting "From Domain"
There were different responses from different softphones:

1) 
Linphone
![linp1](https://user-images.githubusercontent.com/62410690/236880924-6c311836-92ca-46d1-8340-e4ac8e8d2c13.jpg)
Sipnetic
![sipn1](https://user-images.githubusercontent.com/62410690/236880797-355f10e5-d1e8-46c6-aba2-e364175b37d7.jpg)
Zoiper

2) L

**Comment by osering** [comment]
What concerning "@", it is the same test, just instead of ">" I tried character "@", so the URI could potentially look like "ussd@1@olo-tst". It delivered just "ussd@1" part. Have not tried with more than one "@" (before, after).
All these experiments was done with simple From header, containing just simple/compact URI: a@b

----

Strictly according to RFC3261 (SIP), Section 20.20 "From", which refers to Section 20.10 "Contact", URI can be also like this
From: "A. G. Bell" sip:agb@bell-telephone.com".

I formed it 2 ways (with UTF-8 characters (in Latvian you need them and it is not forbidden by RFC):

   [1] From: "dongle1 aāuū" <ussd-1@olo-tst>  (in practice you not always prefix with sip:)
   [2] From: "dongle1 aāuū" <sip:ussd-1@olo-tst>

    domain @olo-tst was attached by FreePBX PJSIP trunk (Advanced) setting "From Domain"
    There were different responses from different softphones:

RESULTS:
[1] From: "dongle1 aāuū" <ussd-1@olo-tst>
Linphone
![linp1](https://user-images.githubusercontent.com/62410690/236881300-e991241a-8e57-4a30-a4fe-ee78d3536651.jpg)

Sipnetic
![sipn1](https://user-images.githubusercontent.com/62410690/236881399-cf428d47-b742-49ec-b4ea-f039cd0a81e2.jpg)

Zoiper
![zoip1](https://user-images.githubusercontent.com/62410690/236881492-b431a9ad-8e4d-427e-ac45-0b0b92e8232e.jpg)

Baresip
![bare1](https://user-images.githubusercontent.com/62410690/236881563-745c9422-d563-4eec-a14a-e4a7621ce537.jpg)

[2] From: "dongle1 aāuū" <sip:ussd-1@olo-tst>
Linphone
![linp2](https://user-images.githubusercontent.com/62410690/236881741-c11a07f3-57d1-480c-b867-05e6b4f918ff.jpg)

Baresip
![bare2](https://user-images.githubusercontent.com/62410690/236881851-7b7414b6-02a3-451c-b617-faab890ff211.jpg)

As can be seen, in Baresip, if the protocol "sip" is skipped, all the characters are present, but are aggressively escaped, both - in Contact and URI (even "<" ">" of URI). Escaped ">" is put before "From Domain", not after (could be the FreePBX fault).
In second (strict accordance with RFC) case, Contact is totally skipped, "<" and ">" skipped, but instead shown protocol name (which is not necessary as it is self-explaining if message is received in SIP client).

**Comment by juha-h** [comment]
Oskars G writes:

> What concerning "@", it is the same test, just instead of ">" I tried
> character "@", so the URI could potentially look like
> ***@***.***@olo-tst". It delivered just ***@***.***" part. Have not tried with
> more than one "@" (before, after).

So was From header like this?

***@***.***;tag=48acfcd7

**Comment by osering** [comment]
I didn't check. If necessary, I can make debugging session.

**Comment by juha-h** [comment]
It turned that there is bug in baresip lib regarding @ char.  I'll try
to get it fixed.

**Comment by osering** [comment]
Great! At least some gain from our conversation!

**Comment by juha-h** [comment]
Thanks for that.  I still don't like the idea that user is presented an invalid SIP URI, for example,
```
sip:first last@example.com
```

**Comment by osering** [comment]
From: "Kā klājas" \<sip:viens-divi@tris.cetri>
is "validated" by RFC3261 (SIP) standard, Section 20.20 "From", and Section 20.10 "Contact", so no invalid characters in URI :)

**Comment by juha-h** [comment]
Yes, URI sip:[viens-divi@tris.cetri](mailto:viens-divi@tris.cetri) is OK, since it only contains these characters:
```
user             =  1*( unreserved / escaped / user-unreserved )
user-unreserved  =  "&" / "=" / "+" / "$" / "," / ";" / "?" / "/"
unreserved  =  alphanum / mark
mark        =  "-" / "_" / "." / "!" / "~" / "*" / "'"
                     / "(" / ")"
```

**Comment by vhbcet** [comment]
Hi guys, does anyone have any information about what to write to the "extensions_custom.conf" file? :)

**Comment by juha-h** [comment]
This project does not have `extensions_custom.conf` file.
